impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeMetricImpl.java

```diff
@@ -0,0 +1,175 @@
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+
+    List<LabelValue> labelValuesCopy = new ArrayList<LabelValue>(labelValues); // Deep copy.
+    checkArgument(labelKeysSize == labelValuesCopy.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValuesCopy, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+
+    List<LabelValue> labelValuesCopy = new ArrayList<LabelValue>(labelValues); // Deep copy.
+    checkArgument(labelKeysSize == labelValuesCopy.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValuesCopy, obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public synchronized Point getDefaultPoint() {
+    List<LabelValue> labelValues = new ArrayList<LabelValue>(labelKeysSize);
+    Point existingPoint = points.get(labelValues);
+    if (existingPoint != null) {
+      return existingPoint;
+    }
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  private synchronized <T> Point addAndGetPoint(
+      List<LabelValue> labelValues, @Nullable T obj, @Nullable ToDoubleFunction<T> function) {
+
+    Point existingPoint = points.get(labelValues);
+    if (existingPoint != null) {
+      throw new IllegalArgumentException("A different point with the same labels already exists");
+    }
+    // Updating the map of Points happens under a lock to avoid multiple add operations
+    // to happen in the same time.
+    Map<List<LabelValue>, Point> pointsCopy = new HashMap<List<LabelValue>, Point>(points);
+    Point newPoint =
+        function != null
+            ? new PointImpl<T>(labelValues, obj, function)
+            : new PointImpl<T>(labelValues);
+
+    pointsCopy.put(labelValues, newPoint);
+    points = Collections.unmodifiableMap(pointsCopy);
+    return newPoint;
+  }
+
+  private List<TimeSeries> getTimeSeries(Clock clock) {
+    List<TimeSeries> timeSeriesList = new ArrayList<TimeSeries>(points.size());
+
+    for (Map.Entry<List<LabelValue>, Point> pointEntry : points.entrySet()) {
+      timeSeriesList.add(
+          TimeSeries.create(
+              pointEntry.getKey(),
+              Collections.singletonList(
+                  io.opencensus.metrics.export.Point.create(
+                      Value.doubleValue(((PointImpl) pointEntry.getValue()).get()), clock.now())),
+              null));
+    }
+    return timeSeriesList;
+  }
+
+  @Override
+  public Metric getMetric(Clock clock) {
+    return Metric.create(metricDescriptor, getTimeSeries(clock));
+  }
+
+  /** Implementation of {@link io.opencensus.metrics.DoubleGaugeMetric.Point}. */
+  public static final class PointImpl<T> extends Point {
+    private final AtomicDouble value = new AtomicDouble(0);
```

Comment:
Please add a comment about DoubleAdder and LongAdder to be used if we change to java8. Maybe we can add the improvement later by allowing multiple implementations but is good to document that.

Reply:
Added TODO.